lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/cache/EvictableLoadingCache.java

```diff
@@ -0,0 +1,380 @@
+    // is *not* necessary for correctness (read-after-invalidate consistency).
+    private final LoadingCache<Token<K>, V> dataCache;
+
+    private EvictableLoadingCache(EvictableCache<K, Token<K>> tokensCache, LoadingCache<Token<K>, V> dataCache)
+    {
+        this.tokensCache = tokensCache;
+        this.dataCache = dataCache;
+    }
+
+    @Override
+    public V get(K key)
+            throws ExecutionException
+    {
+        Token<K> token = tokensCache.get(key, () -> new Token<>(key));
+        return dataCache.get(token);
+    }
+
+    @Override
+    public ImmutableMap<K, V> getAll(Iterable<? extends K> keys)
+            throws ExecutionException
+    {
+        BiMap<K, Token<K>> keyToToken = HashBiMap.create();
+        for (K key : keys) {
+            // This is not bulk, but is fast local operation
+            keyToToken.put(key, tokensCache.get(key, () -> new Token<>(key)));
+        }
+
+        Map<Token<K>, V> values = dataCache.getAll(keyToToken.values());
+
+        BiMap<Token<K>, K> tokenToKey = keyToToken.inverse();
+        ImmutableMap.Builder<K, V> result = ImmutableMap.builder();
+        for (Map.Entry<Token<K>, V> entry : values.entrySet()) {
+            Token<K> token = entry.getKey();
+
+            // While token.getKey() returns equal key, a caller may expect us to maintain key identity, in case equal keys are still distinguishable.
+            K key = tokenToKey.get(token);
+            checkState(key != null, "No key found for %s in %s when loading %s", token, tokenToKey, keys);
+
+            result.put(key, entry.getValue());
+        }
+        return result.buildOrThrow();
+    }
+
+    @Override
+    public V getIfPresent(Object key)
+    {
+        Token<K> token = tokensCache.getIfPresent(key);
+        if (token == null) {
+            return null;
+        }
+        return dataCache.getIfPresent(token);
+    }
+
+    @Override
+    public void invalidate(Object key)
+    {
+        Token<K> token = tokensCache.getIfPresent(key);
+        // Invalidate irrespective of whether we got a token, to invalidate result of ongoing token load (instantiation).
+        tokensCache.invalidate(key);
+        if (token != null) {
+            dataCache.invalidate(token);
+        }
+    }
+
+    @Override
+    public void invalidateAll(Iterable<?> keys)
+    {
+        ImmutableMap<K, Token<K>> tokens = tokensCache.getAllPresent(keys);
+        tokensCache.invalidateAll(keys);
+        // This is racy with respect to concurrent cache loading. dataCache is not guaranteed not to have entries for keys after invalidateAll(keys) returns.
+        // Note that dataCache invalidation is not necessary for correctness. Still one can expect that cache invalidation releases memory.
+        dataCache.invalidateAll(tokens.values());
+    }
+
+    @Override
+    public void invalidateAll()
+    {
+        Set<K> keys = ImmutableSet.copyOf(tokensCache.asMap().keySet());
```

Comment:
Should we invalidate all entries from `dataCache` that do not are not mentioned in `tokens`? So we release stale entries of `dataCache` in the opportunistic way.

Reply:
right, changing.